File: pickzon-socialMedia/SCIMBO/Database/External/VSWordDetector/VSWordDetector.m
//
//  VSWordDetector.m
//  VSWordDetector
//
//  Created by TheTiger on 05/02/14.
//  Copyright (c) 2014 TheTiger. All rights reserved.
//

#import "VSWordDetector.h"
#import <UIKit/UIKit.h>
@interface VSWordDetector ()

@property (strong, nonatomic) id <VSWordDetectorDelegate> delegate;
@property (strong, nonatomic) NSMutableArray *words;
@property (strong, nonatomic) NSMutableArray *wordAreas;

@end

@implementation VSWordDetector
@synthesize delegate = _delegate;
@synthesize words = _words;
@synthesize wordAreas = _wordAreas;

#pragma mark - Initializaiton
-(id)initWithDelegate:(id<VSWordDetectorDelegate>)delegate
{
    self = [super init];
    if (self)
    {
        self.delegate = delegate;
    }
    
    return self;
}

#pragma mark - Adding Detector on view
-(void)addOnView:(id)view
{
    UITapGestureRecognizer *tapGesture = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tapped:)];
    if ([view isKindOfClass:[UITextView class]])
    {
        [view addGestureRecognizer:tapGesture];
        
        UITextView *textView = (UITextView *)view;
        textView.userInteractionEnabled = YES;
        textView.editable = NO;
        textView.scrollEnabled = NO;
    }
    else if ([view isKindOfClass:[UILabel class]])
    {
        [view addGestureRecognizer:tapGesture];
        
        UILabel *label = (UILabel *)view;
        label.userInteractionEnabled = YES;
    }
}

#pragma mark - Tapped
-(void)tapped:(UIGestureRecognizer *)recognizer
{
    
    @try {
    
    if ([recognizer.view isKindOfClass:[UITextView class]])
    {
        UITextView *textView = (UITextView *)recognizer.view;
        
        NSLayoutManager *layoutManager = textView.layoutManager;
        CGPoint location = [recognizer locationInView:textView];
        location.x -= textView.textContainerInset.left;
        location.y -= textView.textContainerInset.top;
        
        // FIND THE CHARACTER WHICH HAVE BEEN TAPPED
        NSInteger characterIndex = [layoutManager characterIndexForPoint:location inTextContainer:textView.textContainer fractionOfDistanceBetweenInsertionPoints:NULL];
        
        if (characterIndex < textView.textStorage.length)
        {
            NSString *word = [self tappedWordInTextView:textView fromIndex:characterIndex];
            if ([self.delegate respondsToSelector:@selector(wordDetector:detectWord:)])
            {
                [self.delegate wordDetector:self detectWord:word];
            }
        }
    } else if ([recognizer.view isKindOfClass:[UILabel class]]) {
        
        UILabel *label = (UILabel *)recognizer.view;
        CGPoint location = [recognizer locationInView:label];
        
        
        NSTextStorage *textStorage = [[NSTextStorage alloc] initWithAttributedString:label.attributedText];
        NSLayoutManager *layoutManager = [[NSLayoutManager alloc] init];
        [textStorage addLayoutManager:layoutManager];
        
        CGSize size = label.bounds.size;
        NSTextContainer *textContainer = [[NSTextContainer alloc] initWithSize:CGSizeMake(label.frame.size.width, label.frame.size.height)];
        textContainer.lineFragmentPadding = 0;
        textContainer.maximumNumberOfLines = label.numberOfLines;
        textContainer.lineBreakMode = label.lineBreakMode;
        textContainer.size = label.frame.size;
        [layoutManager addTextContainer:textContainer];
        
        
        NSUInteger characterIndex = [layoutManager characterIndexForPoint:location
                                                          inTextContainer:textContainer
                                 fractionOfDistanceBetweenInsertionPoints:NULL];
        
        
        
        NSUInteger glyphIndex = [layoutManager glyphIndexForPoint:location inTextContainer:textContainer fractionOfDistanceThroughGlyph:NULL];
        
        NSUInteger characterIndexx = [layoutManager glyphIndexForCharacterAtIndex:glyphIndex];
        
        if (characterIndex < textStorage.length) {
//            NSRange range = NSMakeRange(characterIndex, 1);
//            NSString *value = [label.attributedText.string substringWithRange:range];
//            NSLog(@"%@, %zd, %zd", value, range.location, range.length);
            NSString *word = [self tappedWordInLabel:label fromIndex:characterIndex];
            if ([self.delegate respondsToSelector:@selector(wordDetector:detectWord:)])
            {
                [self.delegate wordDetector:self detectWord:word];
            }
            
            NSLog(@"word==",word);
        }
    }  else if ([recognizer.view isKindOfClass:[UILabel class]])
    {
        UILabel *label = (UILabel *)recognizer.view;
        CGPoint location = [recognizer locationInView:label];
    
        // GETTING ALL WORDS OF LABEL
        self.words = nil;
      //  NSString *str = [[label text] stringByReplacingOccurrencesOfString:@"\n" withString:@" "];
      // self.words = [[[[label attributedText] string] componentsSeparatedByString:@" "] mutableCopy];
       self.words = [[[[label attributedText] string] componentsSeparatedByString:@" "] mutableCopy];

      //  self.words = [[[[label attributedText] string] componentsSeparatedByCharactersInSet:NSCharacterSet.whitespaceAndNewlineCharacterSet] mutableCopy];

        
        
        self.wordAreas = nil;
        self.wordAreas = [[NSMutableArray alloc] init];
        
        __block CGPoint drawPoint = CGPointMake(0, 0);
        CGRect rect = [label frame];
        CGSize space = [@" " sizeWithFont:label.font constrainedToSize:rect.size lineBreakMode:label.lineBreakMode];
        
        
        

        
        // GETTING AREA OF EACH WORD OF LABEL
        [self.words enumerateObjectsUsingBlock:^(NSString *word, NSUInteger idx, BOOL *stop) {
            
            UIFont *font = [label font];
            CGSize size = [word sizeWithAttributes:font];
            
            if(drawPoint.x + size.width > rect.size.width) {
                drawPoint = CGPointMake(0, drawPoint.y + size.height);
            }
            
            [self.wordAreas addObject:[NSValue valueWithCGRect:CGRectMake(drawPoint.x, drawPoint.y, size.width, size.height)]];
            
            drawPoint = CGPointMake(drawPoint.x + size.width + space.width, drawPoint.y);
        }];

        // NOW FINDING THE WORD OF TAPPED AREA
        [self.wordAreas enumerateObjectsUsingBlock:^(NSValue *obj, NSUInteger idx, BOOL *stop) {
            CGRect area = [obj CGRectValue];
            if (CGRectContainsPoint(area, location)) {
                if([self.delegate respondsToSelector:@selector(wordDetector:detectWord:)]){
                    NSString *word = [self.words objectAtIndex:idx];
                    [self.delegate wordDetector:self detectWord:word];
                }
                *stop = YES;
            }
        }];

        
    }
    } @catch (NSException *exception) {
        NSLog(@"%@", exception.description);
    } @finally {
        NSLog(@"ERROR");
    }
    
}

// ONLY FOR TEXT VIEW
-(NSString *)tappedWordInTextView:(UITextView *)textView fromIndex:(NSUInteger)index
{
    
    @try {
   
    NSMutableString *fString = [[NSMutableString alloc] init];
    NSMutableString *sString = [[NSMutableString alloc] init];
    
    // GET STRING BEFORE TAPPED CHARACTER UNTIL SPACE
    
    for (NSInteger i=index; i>=0; i--)
    {
        unichar character = [textView.text characterAtIndex:i];
        NSInteger asciiValue = [[NSString stringWithFormat:@"%d", character] integerValue];
        if (asciiValue == 32)
        {
            // THIS IS SPACE
            break;
        }
        
        [fString appendFormat:@"%c", character];
    }
    
    // REVERSE fString
    
    NSMutableString *reversedString = [NSMutableString stringWithCapacity:[fString length]];
    
    [fString enumerateSubstringsInRange:NSMakeRange(0,[fString length])
                                options:(NSStringEnumerationReverse | NSStringEnumerationByComposedCharacterSequences)
                             usingBlock:^(NSString *substring, NSRange substringRange, NSRange enclosingRange, BOOL *stop) {
                                 [reversedString appendString:substring];
                             }];
    fString = reversedString;
    
    // GET STRING AFTER TAPPED CHARACTER UNTIL SPACE
    
    for (NSInteger i=index+1; i<textView.text.length; i++)
    {
        unichar character = [textView.text characterAtIndex:i];
        NSInteger asciiValue = [[NSString stringWithFormat:@"%d", character] integerValue];
        if (asciiValue == 32)
        {
            // THIS IS SPACE
            break;
        }
        
        [sString appendFormat:@"%c", character];
    }
    
    return [NSString stringWithFormat:@"%@%@", fString, sString];
    
    } @catch (NSException *exception) {
        
    } @finally {
        return @"";
    }
}



// ONLY FOR TEXT VIEW
-(NSString *)tappedWordInLabel:(UILabel *)label fromIndex:(NSUInteger)index
{
    NSMutableString *fString = [[NSMutableString alloc] init];
    NSMutableString *sString = [[NSMutableString alloc] init];
    @try {
    
    // GET STRING BEFORE TAPPED CHARACTER UNTIL SPACE
    
    for (NSInteger i=index; i>=0; i--)
    {
        if ([[label attributedText] length]  > i){
        char character = [label.attributedText.string characterAtIndex:i];
        NSInteger asciiValue = [[NSString stringWithFormat:@"%d", character] integerValue];
        if ((asciiValue == 32) || (asciiValue == 10))
        {
            // THIS IS SPACE
            break;
        }else{
            [fString appendFormat:@"%c", character];
        }
        }
    }
    
    // REVERSE fString
    
    NSMutableString *reversedString = [NSMutableString stringWithCapacity:[fString length]];
    
    [fString enumerateSubstringsInRange:NSMakeRange(0,[fString length])
                                options:(NSStringEnumerationReverse | NSStringEnumerationByComposedCharacterSequences)
                             usingBlock:^(NSString *substring, NSRange substringRange, NSRange enclosingRange, BOOL *stop) {
                                 [reversedString appendString:substring];
                             }];
    fString = reversedString;
    
        NSLog(@"reversedString ==%@",reversedString);
    // GET STRING AFTER TAPPED CHARACTER UNTIL SPACE
    
    for (NSInteger i=index+1; i<label.attributedText.length; i++)
    {
        if ([[label attributedText] length]  > i){
            
            char character = [label.attributedText.string characterAtIndex:i];
            NSInteger asciiValue = [[NSString stringWithFormat:@"%d", character] integerValue];
            if ((asciiValue == 32) || (asciiValue == 10))
            {
                // THIS IS SPACE
                break;
            }else{
                [sString appendFormat:@"%c", character];
            }
        }
        
        
        
        
    }
    
        NSLog(@"fString ==%@",fString);

    return [NSString stringWithFormat:@"%@%@", fString, sString];
    } @catch (NSException *exception) {
        
    } @finally {
        return [NSString stringWithFormat:@"%@%@", fString, sString];

        return  @"";
    }
}




@end

File: pickzon-socialMedia/SCIMBO/Database/External/VSWordDetector/VSWordDetector.swift
//
//  VSWordDetector.swift
//  SCIMBO
//
//  Created by Radheshyam Yadav on 4/11/23.
//  Copyright © 2023 Pickzon Inc. All rights reserved.
//

import UIKit



protocol VSWordDetectorDelegate: AnyObject {
    func wordDetector(wordDetector:VSWordDetector,word:String)
}

class VSWordDetector: NSObject {
    
    var delegate:VSWordDetectorDelegate?
    
    
    //MARK:  - Initializaiton    
    override init() {
        super.init()
    }
    
    
    //MARK: - Adding Detector on view
    
    func addOnView(view:UIView){
        
        let tapGesture = UITapGestureRecognizer(target: self, action: #selector(labelTapped(_:)))
        view.addGestureRecognizer(tapGesture)
        
        if let label = view as? UILabel{
            label.isUserInteractionEnabled = true
        }
    }
    
    
    
    
    @objc func labelTapped(_ sender: UITapGestureRecognizer) {
        
        
        guard let label = sender.view as? ExpandableLabel , sender.state == .ended  else{
            return
        }
        
        guard let text = label.attributedText else { return }
        
        // Get tap location
        let location = sender.location(in: label)

        // Find character index at tap location
        if let characterIndex = label.characterIndex(at: location) {
            
            // Find range of tapped word
            var start = characterIndex
            var end = characterIndex
            
            // Find start of the word
            while start > 0 {
                let range = NSRange(location: start-1, length: 1)
                let substring = (text.string as NSString).substring(with: range)
                if substring == " " || substring == "\n" || substring == "..." || substring == ","{
                    break
                }
                start -= 1
            }
            
            // Find end of the word
            while end < text.length {
                let range = NSRange(location: end, length: 1)
                let substring = (text.string as NSString).substring(with: range)
                if substring == " " || substring == "\n" || substring == "..." || substring == "," || substring == "#" || substring == "@"{
                    break
                }
                end += 1
            }
            
            // Extract tapped word from attributed text
            let range = NSRange(location: start, length: end - start)
            let tappedWord = (text.string as NSString).substring(with: range)
            self.delegate?.wordDetector(wordDetector: self, word: tappedWord)

            // Do something with the tapped word
            print("Tapped word: \(tappedWord)")
        }else{
            self.delegate?.wordDetector(wordDetector: self, word: "")
        }
    }

    
    @objc func handleTapGesture(_ tapGesture: UITapGestureRecognizer) {
        
        guard let label = tapGesture.view as? UILabel else{
            return
        }
        
        guard let text = label.attributedText else { return }
        // Get tap location
                let location = tapGesture.location(in: label)
                
                // Find character index at tap location
                if let characterIndex = label.characterIndex(at: location) {
                    // Find range of word containing tapped character
                    let range = (text.string.encodeEmoji() as NSString).rangeOfWord(at: characterIndex)
                    
                    // Do something with the tapped word
                    let tappedWord = (text.string.encodeEmoji() as NSString).substring(with: range)
                    self.delegate?.wordDetector(wordDetector: self, word: tappedWord)

                    print("Tapped word: \(tappedWord)")
                }
        
      /*  guard let attributedText = label.attributedText else { return }
        
        let layoutManager = NSLayoutManager()
        let textContainer = NSTextContainer(size: label.bounds.size)
        let textStorage = NSTextStorage(attributedString: attributedText)
        
        layoutManager.addTextContainer(textContainer)
        textStorage.addLayoutManager(layoutManager)
        
        let locationOfTouchInLabel = tapGesture.location(in: label)
        let labelSize = label.bounds.size
        let textBoundingBox = layoutManager.usedRect(for: textContainer)
        
        let textContainerOffset = CGPoint(x: (labelSize.width - textBoundingBox.size.width) * 0.5 - textBoundingBox.origin.x,
                                          y: (labelSize.height - textBoundingBox.size.height) * 0.5 - textBoundingBox.origin.y);
        let locationOfTouchInTextContainer = CGPoint(x: locationOfTouchInLabel.x - textContainerOffset.x,
                                                     y: locationOfTouchInLabel.y - textContainerOffset.y);
        
        let indexOfCharacter = layoutManager.characterIndex(for: locationOfTouchInTextContainer, in: textContainer, fractionOfDistanceBetweenInsertionPoints: nil)
        
        if let tappedWordRange = self.getWordRange(at: indexOfCharacter, in: attributedText.string) {
            let tappedWord = (attributedText.string as NSString).substring(with: tappedWordRange)
            self.delegate?.wordDetector(wordDetector: self, word: tappedWord)
            // label.onWordTapped?(tappedWord)
        }*/
    }
    
//    private func getWordRange(at index: Int, in text: String) -> NSRange? {
//        let startIndex = text.index(text.startIndex, offsetBy: index)
//
//        if let endIndex = text.index(startIndex, offsetBy: text.length, limitedBy: text.endIndex) {
//
//            let endIndex = text.index(startIndex, until: { !CharacterSet.alphanumerics.contains($0.unicodeScalars.first!) })
//
//            let range = startIndex..<endIndex
//            return NSRange(range, in: text)
//        }
//    }
    
}



extension UILabel {
    func characterIndex(at point: CGPoint) -> Int? {
        guard let attributedText = attributedText else { return nil }
        
        let textStorage = NSTextStorage(attributedString: attributedText)
        let layoutManager = NSLayoutManager()
        textStorage.addLayoutManager(layoutManager)
        
        let textContainer = NSTextContainer(size: frame.size)
        textContainer.lineFragmentPadding = 0
        //textContainer.maximumNumberOfLines = numberOfLines
        textContainer.lineBreakMode = .byTruncatingTail
        
        layoutManager.addTextContainer(textContainer)
        
        let index = layoutManager.characterIndex(for: point, in: textContainer, fractionOfDistanceBetweenInsertionPoints: nil)
        
        
//        let glyphIndex = layoutManager.glyphIndex(for: point, in: textContainer, fractionOfDistanceThroughGlyph: nil)
//        let characterIndexx = layoutManager.glyphIndexForCharacter(at: glyphIndex)
//
//        return characterIndexx

        
        return index
    }
}


    extension NSString {
        func rangeOfWord(at index: Int) -> NSRange {
            let stringRange = NSRange(location: 0, length: length)
            let options = NSString.EnumerationOptions.byWords.union(.reverse)
            
            var wordRange = NSRange(location: index, length: 0)
            enumerateSubstrings(in: stringRange, options: options) { (substring, substringRange, _, stop) in
                if substringRange.location + substringRange.length <= index {
                    stop.pointee = true
                } else {
                    wordRange = substringRange
                    stop.pointee = true
                }
            }
            return wordRange
        }
    }



extension String {
    func encodeEmoji() -> String {
        let data = self.data(using: .nonLossyASCII, allowLossyConversion: false)!
        return String(data: data, encoding: .utf8)!
    }

    func decodeEmoji() -> String? {
        let data = self.data(using: .utf8)!
        return String(data: data, encoding: .nonLossyASCII)
    }
}
